### ISSUE TITLE
Pairs of irreversible transport reactions that should probably be merged into single reversible reactions

### ISSUE THREAD

**Issue by Devlin-Moyer** [SysBioChalmers/Human-GEM#562]
I used this bit of Python code to systematically identify pairs of irreversible transport reactions oriented in opposite directions, and I think most of them should be combined into single reversible reactions. I was only trying to get transport reactions, but this code also happens to have found a few pairs of irreversible NADH- or NADPH-dependent redox reactions that also seem like they should be combined into single reversible reactions (those reactions are now mentioned in separate issues: #578, #579, and #580).

<details>
  <summary>Python Code</summary>

```python
import cobra
import itertools as it

# these functions & class take a GPR and return a frozenset of
# frozensets, where each frozenset is a list of genes that can
# independently catalyze the reaction, e.g.
# solve('A') -> {{'A'}}
# solve('A or B') -> {{'A'}, {'B'}}
# solve('A and B') -> {{'A', 'B'}}
# solve('A and (B or C)') -> {{'A', 'B'}, {'A', 'C'}}
# solve('(A and B) or (C and D)') -> {{'A', 'B'}, {'C', 'D'}}
def frozenproduct(first, second):
   for e1 in first:
      for e2 in second:
         yield e1.union(e2)

class BoolBlock:
   ops = ['and', 'or']

   def __init__(self):
      self.op = None
      self.possible = set()

   def settype(self, token):
      assert token in BoolBlock.ops
      if self.op is None:
         self.op = token
      else:
         assert self.op == token

   def add(self, new):
      if self.op == 'and':
         self.possible = set(frozenproduct(self.possible, new))
      else:
         self.possible.update(new)

def tokenize(string):
   buffer = ''
   for c in string:
      if c in ' ()':
         if buffer.strip():
            yield buffer.strip()
         if c.strip():
            yield c
         buffer = ''
      else:
         buffer += c

   if buffer.strip():
      yield buffer.strip()

def recurse(token_stream):
   current = BoolBlock()
   for token in token_stream:
      if token == '(':
         current.add(recurse(token_stream).possible)
      elif token == ')':
         return current
      elif token in BoolBlock.ops:
         current.settype(token)
      else:
         gene = {frozenset({token,}), }
         current.add(gene)
   return current

def solve(string):
   return recurse(tokenize(string)).possible

model = cobra.io.load_yaml_model('Human-GEM.yml')

bad_pairs = set()
# loop over metabolites
for ref_met in list(model.metabolites):
    # find reactions that move this metabolite between compartments
    trans_rxns = [r for r in ref_met.reactions if any(
        other_met.id[:-1] == ref_met.id[:-1]
        for other_met in r.metabolites
    )]
    # loop over pairs of those reactions
    for (r1, r2) in it.combinations(trans_rxns, 2):
        # see if the two stoichiometry dicts are either identical or identical
        # if you multiply all coefficients by -1 (i.e. same or opposite
        # directions)
        same_dir = r1.metabolites == r2.metabolites
        opp_dir = r1.metabolites == {
            m : -1 * s for (m,s) in r2.metabolites.items()
        }
        # see if either (solved) GPR is a subset of the other
        r1_gcs = solve(r1.gene_reaction_rule)
        r2_gcs = solve(r2.gene_reaction_rule)
        gpr_1_in_2 = all(g1 in r2_gcs for g1 in r1_gcs)
        gpr_2_in_1 = all(g2 in r1_gcs for g2 in r2_gcs)
        if (same_dir or opp_dir) and (gpr_1_in_2 or gpr_2_in_1):
            # order reactions by ID so we don't add the same pair twice in
            # opposite orders
            if int(r1.id[3:]) < int(r2.id[3:]):
                bad_pairs.add((r1, r2))
            else:
                bad_pairs.add((r2, r1))
```
</details>

These should probably all be collapsed into single, reversible reactions

- [X] MAR02174: 3-oxotetradecanoyl-[ACP] [c] + H+ [c] + NADPH [c] --> HMA [c] + NADP+ [c], GPR: `ENSG00000169710`
MAR02185: HMA [c] + NADP+ [c] --> 3-oxotetradecanoyl-[ACP] [c] + H+ [c] + NADPH [c], GPR: `ENSG00000169710`

- [X] MAR11583: PI pool [c] --> PI pool [e], no GPR
MAR11649: PI pool [e] --> PI pool [c], no GPR

- [X] MAR11728: 2-Hydroxy-Isocaproate [e] --> 2-Hydroxy-Isocaproate [c], no GPR
MAR11729: 2-Hydroxy-Isocaproate [c] --> 2-Hydroxy-Isocaproate [e], no GPR

- [X] MAR10468: 3-Hydroxycinnamic Acid [e] --> 3-Hydroxycinnamic Acid [c], no GPR
MAR10469: 3-Hydroxycinnamic Acid [c] --> 3-Hydroxycinnamic Acid [e], no GPR

- [X] MAR10466: P-Cresol Sulfate [c] --> P-Cresol Sulfate [e], no GPR
MAR10467: P-Cresol Sulfate [e] --> P-Cresol Sulfate [c], no GPR

- [X] MAR10470: 3-Hydroxyphenylpropionic Acid (3-Hppa) [e] --> 3-Hydroxyphenylpropionic Acid (3-Hppa) [c], no GPR
MAR10471: 3-Hydroxyphenylpropionic Acid (3-Hppa) [c] --> 3-Hydroxyphenylpropionic Acid (3-Hppa) [e], no GPR

- [X] MAR01822: vitamin D3 [m] --> vitamin D3 [c], no GPR
MAR01823: vitamin D3 [c] --> vitamin D3 [m], no GPR

- [X] MAR11986: apoB100 [c] --> apoB100 [e], no GPR
MAR11988: apoB100 [e] --> apoB100 [c], no GPR

- [X] MAR11989: apoC1 [c] --> apoC1 [e], no GPR
MAR11991: apoC1 [e] --> apoC1 [c], no GPR

- [X] MAR11992: apoC2 [c] --> apoC2 [e], no GPR
MAR11994: apoC2 [e] --> apoC2 [c], no GPR

- [X] MAR11995: apoC3 [c] --> apoC3 [e], no GPR
MAR11997: apoC3 [e] --> apoC3 [c], no GPR

- [X] MAR11998: apoE [c] --> apoE [e], no GPR
MAR12000: apoE [e] --> apoE [c], no GPR

- [X] MAR11586: 1-Stearoylglycerophosphoinositol [c] --> 1-Stearoylglycerophosphoinositol [e], no GPR
MAR11652: 1-Stearoylglycerophosphoinositol [e] --> 1-Stearoylglycerophosphoinositol [c], no GPR

- [x] MAR11584: 1-Palmitoylglycerophosphoinositol [c] --> 1-Palmitoylglycerophosphoinositol [e], no GPR
MAR11651: 1-Palmitoylglycerophosphoinositol [e] --> 1-Palmitoylglycerophosphoinositol [c], no GPR

- [X] MAR11585: 1-Arachidonoylglycerophosphoinositol [c] --> 1-Arachidonoylglycerophosphoinositol [e], no GPR
MAR11650: 1-Arachidonoylglycerophosphoinositol [e] --> 1-Arachidonoylglycerophosphoinositol [c], no GPR

- [X] MAR11644: Sm (D18:0/24:0), Sphingomyelin [c] --> Sm (D18:0/24:0), Sphingomyelin [e], no GPR
MAR11725: Sm (D18:0/24:0), Sphingomyelin [e] --> Sm (D18:0/24:0), Sphingomyelin [c], no GPR

- [X] MAR11630: Sm (D18:0/24:1), Sphingomyelin [c] --> Sm (D18:0/24:1), Sphingomyelin [e], no GPR
MAR11697: Sm (D18:0/24:1), Sphingomyelin [e] --> Sm (D18:0/24:1), Sphingomyelin [c], no GPR

- [X] MAR11645: Sm (D18:0/25:0), Sphingomyelin [c] --> Sm (D18:0/25:0), Sphingomyelin [e], no GPR
MAR11727: Sm (D18:0/25:0), Sphingomyelin [e] --> Sm (D18:0/25:0), Sphingomyelin [c], no GPR

- [X] MAR11631: Sm (D18:1/14:0), Sphingomyelin [c] --> Sm (D18:1/14:0), Sphingomyelin [e], no GPR
MAR11699: Sm (D18:1/14:0), Sphingomyelin [e] --> Sm (D18:1/14:0), Sphingomyelin [c], no GPR

- [X] MAR11632: Sm (D18:1/15:0), Sphingomyelin [c] --> Sm (D18:1/15:0), Sphingomyelin [e], no GPR
MAR11701: Sm (D18:1/15:0), Sphingomyelin [e] --> Sm (D18:1/15:0), Sphingomyelin [c], no GPR

- [X] MAR11633: Sm (D18:1/16:0), Sphingomyelin [c] --> Sm (D18:1/16:0), Sphingomyelin [e], no GPR
MAR11703: Sm (D18:1/16:0), Sphingomyelin [e] --> Sm (D18:1/16:0), Sphingomyelin [c], no GPR

- [X] MAR11634: Sm (D18:1/16:1), Sphingomyelin [c] --> Sm (D18:1/16:1), Sphingomyelin [e], no GPR
MAR11705: Sm (D18:1/16:1), Sphingomyelin [e] --> Sm (D18:1/16:1), Sphingomyelin [c], no GPR

- [X] MAR11635: Sm (D18:1/17:0), Sphingomyelin [c] --> Sm (D18:1/17:0), Sphingomyelin [e], no GPR
MAR11707: Sm (D18:1/17:0), Sphingomyelin [e] --> Sm (D18:1/17:0), Sphingomyelin [c], no GPR

- [X] MAR11636: Sm (D18:1/18:0), Sphingomyelin [c] --> Sm (D18:1/18:0), Sphingomyelin [e], no GPR
MAR11709: Sm (D18:1/18:0), Sphingomyelin [e] --> Sm (D18:1/18:0), Sphingomyelin [c], no GPR

- [X] MAR11637: Sm (D18:1/18:1), Sphingomyelin [c] --> Sm (D18:1/18:1), Sphingomyelin [e], no GPR
MAR11711: Sm (D18:1/18:1), Sphingomyelin [e] --> Sm (D18:1/18:1), Sphingomyelin [c], no GPR

- [X] MAR11638: Sm (D18:1/20:0), Sphingomyelin [c] --> Sm (D18:1/20:0), Sphingomyelin [e], no GPR
MAR11713: Sm (D18:1/20:0), Sphingomyelin [e] --> Sm (D18:1/20:0), Sphingomyelin [c], no GPR

- [X] MAR11639: Sm (D18:1/20:1), Sphingomyelin [c] --> Sm (D18:1/20:1), Sphingomyelin [e], no GPR
MAR11715: Sm (D18:1/20:1), Sphingomyelin [e] --> Sm (D18:1/20:1), Sphingomyelin [c], no GPR

- [X] MAR11640: Sm (D18:1/21:0), Sphingomyelin [c] --> Sm (D18:1/21:0), Sphingomyelin [e], no GPR
MAR11717: Sm (D18:1/21:0), Sphingomyelin [e] --> Sm (D18:1/21:0), Sphingomyelin [c], no GPR

- [X] MAR11641: Sm (D18:1/22:0), Sphingomyelin [c] --> Sm (D18:1/22:0), Sphingomyelin [e], no GPR
MAR11719: Sm (D18:1/22:0), Sphingomyelin [e] --> Sm (D18:1/22:0), Sphingomyelin [c], no GPR

- [X] MAR11642: Sm (D18:1/22:1), Sphingomyelin [c] --> Sm (D18:1/22:1), Sphingomyelin [e], no GPR
MAR11721: Sm (D18:1/22:1), Sphingomyelin [e] --> Sm (D18:1/22:1), Sphingomyelin [c], no GPR

- [X] MAR11643: Sm (D18:1/23:0), Sphingomyelin [c] --> Sm (D18:1/23:0), Sphingomyelin [e], no GPR
MAR11723: Sm (D18:1/23:0), Sphingomyelin [e] --> Sm (D18:1/23:0), Sphingomyelin [c], no GPR

**Comment by Devlin-Moyer** [comment]
@haowang-bioinfo I just updated this comment; I wrote some Python code to systematically identify all reactions like this and significantly expanded the list I happened to come across earlier

**Comment by haowang-bioinfo** [comment]
@Devlin-Moyer this is very good!

again, could you please split this issue into 4 or 5 ones, by moving the subsections to individual issues and each become more specific and easy to solve. Aggregated complex issues often delay things from getting better and become error-prone while fixing

**Comment by feiranl** [comment]
I am ok with the merge as long as also merge all cross reference annotations and rxnNotes.

**Comment by haowang-bioinfo** [comment]
it might be worth to note that transporters may function in irreversible way, such as in "import" or "export" mode

**Comment by Devlin-Moyer** [comment]
while that may be the case in general, I felt comfortable proposing that these specific transport reactions be merged, since all but two have no genes associated with them, and the only two that do (`MAR02174` and `MAR02185`) have exactly the same gene associated with them (`ENSG00000169710`)

**Comment by haowang-bioinfo** [comment]
fixed in #625
